### DISCUSSION TITLE
Cannot pass lambda expression as Button's `on_click()` function

### DISCUSSION THREAD

**Discussion by georgjz** [ArthurSonzogni/FTXUI#323] [Q&A] [answered]
I'm not sure if this is a bug or just me being dense (relearning C++, last time I touched it was pre-C++11), but it seems I can't pass a lambda as Button's `on_click()` argument:

```cpp
    Component exit_button = Button( "Exit", screen.ExitLoopClosure() );
    Component exit_button = Button( "Exit", [&screen]{ screen.ExitLoopClosure(); } );
```

The first example will work, while the latter won't. I've looked into `FTXUI/src/ftxui/component/button.cpp` and I can't see a reason why it shouldn't be able to call a lambda expression for `on_click()` (again, relearning C++ currently, could be wrong).

Would it some cases make code easier to write I could pass lambdas directly.

**Comment by ArthurSonzogni** [comment, marked as answer]
Hello,

`ScreenInteractive::ExitLoopClosure()` returns the closure. It does not execute it.

So you should use
```cpp
    Component exit_button = Button( "Exit", screen.ExitLoopClosure());
```

or:

```cpp
    Component exit_button = Button( "Exit", [&screen] { screen.ExitLoopClosure()(); });
```

**Reply by ArthurSonzogni** [reply]
To avoid people being confused, it might worth to expose:
```cpp
void ScreenInteractive::ExitLoop() {
    quit_ = true;
    event_sender_.reset();
}
```

And rewrite ScreenInteractive::ExitLoopClosure() as:
```cpp
ScreenInteractive::Callback ScreenInteractive::ExitLoopClosure() {
  return [this] { Exit(); }
}
```

I would be happy to add this if you would like to make a PR.

**Reply by georgjz** [reply]
Thanks for the clarification! I didn't think to look at `ExitLoopClosure()`, but that makes a lot of sense. I'll consider a PR this weekend when I have more times.

**Comment by kcbit** [comment]
Hello,
I having problem with  screen_.ExitLoopClosure();
in the main 
```cpp
  auto screen = ftxui::ScreenInteractive::FitComponent();
    Cgui cgui(screen);
    cgui.showdialoge();
    std::cout << "Delete keys " << cgui.selected_item << "\n";
```

and in Cgui
```cpp
Cgui::Cgui(ftxui::ScreenInteractive &screen) : selected_index(0), screen_(screen) {
}

void Cgui::showdialoge() {
    std::vector<std::string> key = {
            "DKLK4PM.TEST.WPG",
            "DKLK4XAC.WEST",
            "DKLK4PM.TEST.WEST",
    };
    auto button_del = ftxui::Button("Keys", [&key, this] {
        Selected_key(key[selected_index]);
    });
    screen_.Loop(ftxui::Container::Vertical({
           ftxui::Dropdown(&key, &selected_index),
           button_del,
   }));
}
void Cgui::Selected_key(std::string &key) {
    selected_item = key;
    screen_.ExitLoopClosure();
    // want to return to main, do not work, why?
}
```
appreciate any help.
kcbit

**Reply by ArthurSonzogni** [reply]
@kcbit
We were saying above:
```cpp
screen_.ExitLoopClosure();
```
returns a function. You need to call it.
```cpp
screen_.ExitLoopClosure()();
```

**Comment by kcbit** [comment]
I am very grateful for your help.
kcbit.
